platform/cog-platform-drm.c

```diff
@@ -682,6 +708,91 @@ drm_create_buffer_for_bo (struct gbm_bo *bo, struct wl_resource *buffer_resource
+    if (format != WL_SHM_FORMAT_ARGB8888 && format != WL_SHM_FORMAT_XRGB8888) {
+        g_warning ("failed to handle non-32-bit ARGB/XRGB format");
+        return NULL;
+    }
+
+    int32_t width = wl_shm_buffer_get_width (shm_buffer);
+    int32_t height = wl_shm_buffer_get_height (shm_buffer);
+
+    // TODO: don't ignore the alpha channel in case of ARGB8888 SHM data
+    uint32_t gbm_format = GBM_FORMAT_XRGB8888;
+    struct gbm_bo *bo = gbm_bo_create (gbm_data.device, width, height, gbm_format,
+                                       GBM_BO_USE_SCANOUT | GBM_BO_USE_WRITE);
+    if (!bo) {
+        g_warning ("failed to create a gbm_bo object");
+        return NULL;
+    }
+
+    uint32_t in_handles[4] = { 0, };
+    uint32_t in_strides[4] = { 0, };
+    uint32_t in_offsets[4] = { 0, };
+    in_handles[0] = gbm_bo_get_handle (bo).u32;
+    in_strides[0] = gbm_bo_get_stride (bo);
+
+    uint32_t fb_id = 0;
+    int ret = drmModeAddFB2 (drm_data.fd, width, height, gbm_format,
+                             in_handles, in_strides, in_offsets,
+                             &fb_id, 0);
+    if (ret) {
+        gbm_bo_destroy (bo);
+        g_warning ("failed to create framebuffer: %s", strerror (errno));
+        return NULL;
+    }
+
+    struct buffer_object *buffer = g_new0 (struct buffer_object, 1);
+    wl_list_insert (&drm_data.buffer_list, &buffer->link);
+    buffer->destroy_listener.notify = destroy_buffer_notify;
+    wl_resource_add_destroy_listener (buffer_resource, &buffer->destroy_listener);
+
+    buffer->fb_id = fb_id;
+    buffer->bo = bo;
+    buffer->buffer_resource = buffer_resource;
+
+    return buffer;
+}
+
+static void
+drm_copy_shm_buffer_into_bo (struct wl_shm_buffer *shm_buffer, struct gbm_bo *bo)
+{
+    int32_t width = wl_shm_buffer_get_width (shm_buffer);
+    int32_t height = wl_shm_buffer_get_height (shm_buffer);
+    int32_t stride = wl_shm_buffer_get_stride (shm_buffer);
+
+    uint32_t bo_stride = 0;
+    void *map_data = NULL;
+    gbm_bo_map (bo, 0, 0, width, height, GBM_BO_TRANSFER_WRITE, &bo_stride, &map_data);
+    if (!map_data)
+        return;
+
+    wl_shm_buffer_begin_access (shm_buffer);
+
+    uint8_t *src = wl_shm_buffer_get_data (shm_buffer);
+    uint8_t *dst = map_data;
+
+    uint32_t bo_width = gbm_bo_get_width (bo);
+    uint32_t bo_height = gbm_bo_get_height (bo);
+    if (!(width == bo_width && height == bo_height && stride == bo_stride)) {
+        for (uint32_t y = 0; y < height; ++y) {
+            for (uint32_t x = 0; x < width; ++x) {
+                dst[bo_stride * y + 4 * x + 0] = src[stride * y + 4 * x + 0];
+                dst[bo_stride * y + 4 * x + 1] = src[stride * y + 4 * x + 1];
+                dst[bo_stride * y + 4 * x + 2] = src[stride * y + 4 * x + 2];
+                dst[bo_stride * y + 4 * x + 3] = src[stride * y + 4 * x + 3];
+            }
+        }
+    } else
+        memcpy(dst, src, stride * height);
```

Comment:
Nit: missing space between function name and parenthesis:

```suggestion
        memcpy (dst, src, stride * height);
```